src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+            }
+
+            /* reset transmitter state */
+            tx_state = CS8900_TX_IDLE;
+    
+#ifdef CS8900_DEBUG_RXTX_STATE
+            log_message(cs8900_log, "TX: sent  frame (length=%04x)",tx_length);
+#endif
+
+            /* reset tx status */
+            cs8900_set_tx_status(0,0);
+        }
+    }
+}
+
+u8 cs8900a_device::cs8900_read_rx_buffer(int odd_address)
+{
+    if(rx_state != CS8900_RX_GOT_FRAME) {
+#ifdef CS8900_DEBUG_WARN_RXTX
+        log_message(cs8900_log, "WARNING! RX Read without frame available! (odd=%d)",
+                    odd_address);
+#endif
+        /* always reads zero on HW */
+        return 0;
+    } else {
+        /*
+         According to the CS8900 spec, the handling is the following:
+         first read H, then L (RX_STATUS), then H, then L (RX_LENGTH).
+         Inside the RX frame data, we always get L then H, until the end is reached.
+
+                                     even    odd
+         CS8900_PP_ADDR_RXSTATUS:         -     proceed
+         CS8900_PP_ADDR_RXLENGTH:         -     proceed 
+         CS8900_PP_ADDR_RX_FRAMELOC:      -       -
+         CS8900_PP_ADDR_RX_FRAMELOC+2: proceed    -
+         CS8900_PP_ADDR_RX_FRAMELOC+4: proceed    -
+
+         */
+        u16 addr = odd_address ? 1:0; 
+        u8 value;
+        /* read RXSTATUS or RX_LENGTH */
+        if(rx_count<4) {
+            addr  += rx_buffer;
+            value  = GET_PP_8(addr);
+            rx_count++;
+
+            /* incr after RXSTATUS or RX_LENGTH even (L) read */
+            if(!odd_address)
+                rx_buffer += 2;
+        } 
+        /* read frame data */
+        else {
+            /* incr before frame read (but not in first word) */
+            if((rx_count>=6) && (!odd_address))
+                rx_buffer += 2;
+
+            addr += rx_buffer;
+            value = GET_PP_8(addr);
+            rx_count++;
+        }
+        
+#ifdef CS8900_DEBUG_RXTX_DATA
+        log_message(cs8900_log,"RX: %04x/%04x: %02x (buffer=%04x,odd=%d)",
+                    rx_count,rx_length+4,value,addr,odd_address);
+#endif
+
+        /* check frame end */
+        if(rx_count>=rx_length+4) {
+            /* reset receiver state to idle */
+            rx_state = CS8900_RX_IDLE;
+#ifdef CS8900_DEBUG_RXTX_STATE
+            log_message(cs8900_log,"RX: read  frame (length=%04x)",rx_length);
+#endif
+        }        
+        return value;
+    }
+}
+
+/* ------------------------------------------------------------------------- */
+/* handle side-effects of read and write operations */
+
+/*
+ This is called *after* the relevant octets are written
+*/
+void cs8900a_device::cs8900_sideeffects_write_pp(u16 ppaddress, int odd_address)
+{
+    u16 content = GET_PP_16( ppaddress );
+
+    assert((ppaddress & 1) == 0);
+
+    switch (ppaddress)
+    {
+    case CS8900_PP_ADDR_CC_RXCFG:
+        /* Skip_1 Flag: remove current (partial) tx frame and restore state */
+        if (content & 0x40) {
```

Comment:
Please be consistent with your style.  You seem to mostly use no space between keywords and their controlling expressions (e.g. `if(...`) but here you have an intervening space (`switch (...` and `if (...`).